File: 프로그래머스/2/388352. 비밀 코드 해독/비밀 코드 해독.cpp
#include <string>
#include <vector>
#include <set>

using namespace std;

int answer = 0;
int n;
vector<vector<int>> q;
vector<int> ans;

bool checkPassword(vector<int> number) {
    
    set<int> numSet;
    for (int num : number) {
        numSet.insert(num);
    }
    
    for (int i = 0; i < q.size(); i++) {
        int count = 0;
        for (int num : q[i]) {
            if (numSet.count(num) > 0) {
                count++;
            }
        }
        
        if (count != ans[i]) {
            return false;
        }
    }
    return true;
}

void combinateAll(int index, int curr, vector<int> number) {
    
    if (index == 5) {
        if (checkPassword(number)) {
            answer++;
        }
        return;
    }
    
    for (int i = curr; i <= n; i++) {
        number[index] = i;
        combinateAll(index + 1, i + 1, number);
    }
}

int solution(int n, vector<vector<int>> q, vector<int> ans) {
    ::n = n;
    ::q = q;
    ::ans = ans;
    
    combinateAll(0, 1, vector<int>(5));
    
    return answer;
}

File: 프로그래머스/2/388352. 비밀 코드 해독/비밀 코드 해독.java
import java.util.*;

class Solution {
    
    int n;
    int[][] q;
    int[] ans;
    int answer = 0;
    
    public int solution(int n, int[][] q, int[] ans) {
        this.n = n;
        this.q = q;
        this.ans = ans;
        
        combinateAll(1, 0, new int[5]);
        
        return answer;
    }
    
    private void combinateAll(int curr, int index, int[] number) {
        
        if (index == 5) {
            // 검증
            if (checkPassword(number)) {
                answer++;
            }
            return;
        }
        
        for (int i = curr; i <= n; i++) {
            number[index] = i;
            combinateAll(i + 1, index + 1, number);
        }
    }
    
    private boolean checkPassword(int[] number) {
        
        Set<Integer> set = new HashSet<>();
        for (int num : number) {
            set.add(num);
        }
        
        for (int i = 0; i < q.length; i++) {
            
            int count = 0;
            for (int num : q[i]) {
                if (set.contains(num)) {
                    count++;
                }
            }
            
            if (count != ans[i]) {
                return false;
            }
        }
        
        return true;
    }
}